### PR TITLE
Add download from gstorage and cluster_config flag

### PR DESCRIPTION
**Pull request by werkaaa**
Add download from gstorage and cluster_config flag

- cluster_config flag for multi-worker training.
- Downloading configuration flags from gstorage.
- Handling gstorage config paths.
- Tests for handling gstorage config paths.
- Small strategy update, I haven't set everything correctly last time.

### CODE REVIEW

File: ddsp/training/ddsp_run.py [hunk 1]
```diff
@@ -112,6 +119,58 @@ def delay_start():
...
+  Args:
+    paths: Single path or a list of paths.
+
+  Returns:
+    Single local path or a list of local paths.
+  """
+  if isinstance(paths, str):
+    if re.match('gs://*', paths):
+      local_name = os.path.basename(paths)
+      download_from_gstorage(paths, os.path.join(GIN_PATH, local_name))
```
Comment (jesseengel) [inline ddsp/training/ddsp_run.py:159]: GIN_PATH should be passed in as an arg, not a global.

File: ddsp/training/ddsp_run_test.py [hunk 2]
```diff
@@ -0,0 +1,78 @@
+# Copyright 2020 The DDSP Authors.
```
Comment (jesseengel) [inline ddsp/training/ddsp_run_test.py:1]: this should be `ddsp/training/cloud_test.py`

File: ddsp/training/ddsp_run.py [hunk 3]
```diff
@@ -112,6 +119,58 @@ def delay_start():
     logging.info('Waiting for %i second(s)', delay_time)
     time.sleep(delay_time)
 
+def download_from_gstorage(gstorage_path, local_path):
+  """Downloads the bucket.
```
[the remark this answers is not in the document]
Comment (werkaaa) [reply ddsp/training/ddsp_run.py:123]: Yes, you are right, I missed it. Thanks!
Done.

File: ddsp/training/ddsp_run.py [hunk 4]
```diff
@@ -112,6 +119,58 @@ def delay_start():
...
+
+  Args:
+    gstorage_path: Path to the object inside the bucket that needs to be
```
[the remark this answers is not in the document]
Comment (werkaaa) [reply ddsp/training/ddsp_run.py:126]: Okay, I wanted to go with [documentation](https://cloud.google.com/storage/docs/downloading-objects#storage-download-object-python) naming, but in this context "file" may fit better.
Done.

File: ddsp/training/ddsp_run.py [hunk 5]
```diff
@@ -112,6 +119,58 @@ def delay_start():
...
+  blob_name = os.path.relpath(gstorage_path, bucket_name)
+
+  storage_client = storage.Client()
+
+  bucket = storage_client.bucket(bucket_name)
+  blob = bucket.blob(blob_name)
+
+  blob.download_to_filename(local_path)
+  logging.info(
+      'Downloaded the bucket object %s. Current location: %s',
```
[the remark this answers is not in the document]
Comment (werkaaa) [reply ddsp/training/ddsp_run.py:141]: Done.

File: ddsp/training/ddsp_run.py [hunk 6]
```diff
@@ -112,6 +119,58 @@ def delay_start():
     logging.info('Waiting for %i second(s)', delay_time)
     time.sleep(delay_time)
 
+def download_from_gstorage(gstorage_path, local_path):
```
[the remark this answers is not in the document]
Comment (werkaaa) [reply ddsp/training/ddsp_run.py:122]: Done.

File: ddsp/training/ddsp_run.py [hunk 7]
```diff
@@ -112,6 +119,58 @@ def delay_start():
...
+
+  bucket = storage_client.bucket(bucket_name)
+  blob = bucket.blob(blob_name)
+
+  blob.download_to_filename(local_path)
+  logging.info(
+      'Downloaded the bucket object %s. Current location: %s',
+      gstorage_path, local_path)
+
+def handle_gstorage_paths(paths):
```
[the remark this answers is not in the document]
Comment (werkaaa) [reply ddsp/training/ddsp_run.py:144]: Done.

File: ddsp/training/ddsp_run.py (hunk 1 above)
Comment (werkaaa) [reply ddsp/training/ddsp_run.py:159]: Done.

File: ddsp/training/ddsp_run_test.py (hunk 2 above)
Comment (werkaaa) [reply ddsp/training/ddsp_run_test.py:1]: Done.

**Comment by jesseengel** [comment]
Cool, I think you need to sync your branch to head.

**Comment by werkaaa** [comment]
> Cool, I think you need to sync your branch to head.

Yes, you're right. I've synced with head.

**Comment by werkaaa** [comment]
Hi! I've added early_stop_loss_value flag in ddsp_run.py. I has been using this parameter in train_util.train through Gin config flag but while writing the documentation for the final PR I thought it should be also accessible for the user as a normal flag. Hope this is the last change I make before PR with the Dockerfile and configs.

**Comment by jesseengel** [comment]
Hmm, seems this PR has already been merged, can you make a new one with the FLAG changes?

**Comment by werkaaa** [comment]
> Hmm, seems this PR has already been merged, can you make a new one with the FLAG changes?

Sure, will do :)

**Pull request closed without merging**